src/libsystemd/sd-event/sd-event.c

```diff
@@ -2819,17 +2819,15 @@ static int event_source_enter_ratelimited(sd_event_source *s) {
         if (EVENT_SOURCE_IS_TIME(s->type))
                 event_source_time_prioq_remove(s, event_get_clock_data(s->event, s->type));
 
-        /* Now, let's add the event source to the monotonic clock instead */
-        r = event_source_time_prioq_put(s, &s->event->monotonic);
+        /* And let's take the event source officially offline */
+        r = event_source_offline(s, s->enabled, /* ratelimited= */ true);
         if (r < 0)
                 goto fail;
 
-        /* And let's take the event source officially offline */
-        r = event_source_offline(s, s->enabled, /* ratelimited= */ true);
-        if (r < 0) {
-                event_source_time_prioq_remove(s, &s->event->monotonic);
+        /* Now, let's add the event source to the monotonic clock instead */
+        r = event_source_time_prioq_put(s, &s->event->monotonic);
+        if (r < 0)
                 goto fail;
-        }
```

Comment:
oh, ha, i think you indeed found the issue here.

Hmm, I wonder if a more robust fix for future changes might be to change event_source_offline() and event_source_online() to call event_source_time_prioq_reshuffle() if in ratelimited mode, because this is what this boils down to: we modify the ordering keys without reshuffling the queue due to that. And I think modifying the keys should always be OK, as long as we then reshuffle things.

Hence, could you rework this so that the two functions offline/online functions call event_source_time_prioq_reshuffle() in any case. And then I guess  event_source_time_prioq_reshuffle()  should be tweaked so that it is safe to call it for any event source. Right now it asserts when called for an event source that is neither a timer event source, nor is ratelimited. It shsould just become a NOP in that case.

Reply:
Sure. OK.